src/metpy/io/text.py

```diff
@@ -0,0 +1,134 @@
+# Copyright (c) 2021 MetPy Developers.
+# Distributed under the terms of the BSD 3-Clause License.
+# SPDX-License-Identifier: BSD-3-Clause
+"""Support reading information from various text file formats."""
+
+import contextlib
+from datetime import datetime
+import re
+import string
+
+import numpy as np
+import pandas as pd
+from shapely.geometry import LineString, Point
+
+from ._tools import open_as_needed
+
+
+def _decode_coords(coordinates):
+    """Turn a string of coordinates from WPC coded surface bulletin into a lon/lat tuple.
+
+    Parameters
+    ----------
+    coordinates : str
+        A string of numbers that can be converted into a lon/lat tuple
+
+    Returns
+    -------
+    (lon, lat) : tuple
+        Longitude and latitude parsed from `coordinates`
+
+    Notes
+    -----
+    In the WPC coded surface bulletin, latitude and longitude are given in degrees north and
+    degrees west, respectively. Therefore, this function always returns latitude as a positive
+    number and longitude as a negative number.
+
+    Examples
+    --------
+    >>> _decode_coords('4731193')
+    (-119.3, 47.3)
+
+    """
+    # Based on the number of digits, find the correct place to split between lat and lon
+    # Hires bulletins provide 7 digits for coordinates; regular bulletins provide 4 or 5 digits
+    split_pos = int(len(coordinates) / 2)
+    lat, lon = coordinates[:split_pos], coordinates[split_pos:]
+
+    # Insert decimal point at the correct place and convert to float
+    lat = float(f'{lat[:2]}.{lat[2:]}')
+    lon = -float(f'{lon[:3]}.{lon[3:]}')
+
+    return lon, lat
+
+
+def _regroup_lines(iterable):
+    starting_num = re.compile('^[0-9]')
+    lines = list(iterable)[::-1]
+    while lines:
+        line = lines.pop()
+        if not line.strip():
+            continue
+        parts = line.split()
+        while lines and starting_num.match(lines[-1]):
+            parts.extend(lines.pop().split())
+        yield parts
+
+
+def parse_wpc_surface_bulletin(bulletin, year=None):
+    """Parse a coded surface bulletin from NWS WPC into a GeoPandas GeoDataFrame.
+
+    Parameters
+    ----------
+    bulletin : str or file-like object
+        If str, the name of the file to be opened. If `bulletin` is a file-like object,
+        this will be read from directly.
+
```

Comment:
We should likely document `year` and its behavior on `None` (including if the header isn't available, I think.)

Still, I'm not super familiar with these WPC bulletins. Can we rely on the header structure well enough that parsing the year (assuming it's there and un-massaged) will usually be a non-issue? 🤞

Reply:
Oops, never documented after I added it so that we could get actual datetimes out.

Most sources of these bulletins are going to have the NWS product header that will have the year and this should "just" work. Strictly speaking, though, the actual product doesn't have it so I assume there has to be some weird case out there. I also wonder if there's some weird corner case where the valid time of the bulletin and issue time of the product (and the header) crosses the year boundary. This would be an escape hatch for that, which I like to have when we're doing more "value-added, heuristic-based parsing" 😆